File: src/Day1/Part2.hs
module Day1.Part2 where

import Control.Monad ( guard )
import Data.List     ( tails )
import Util.Util     ( parseInts )

day1part2 :: String -> String
day1part2 input = show $ a * b * c
  where xs        = parseInts input
        (a, b, c) = head $ solve 2020 xs

solve :: Int -> [Int] -> [(Int, Int, Int)]
solve c xs = do
    (x:ys) <- tails xs
    (y:zs) <- tails ys
    z <- zs
    guard $ x + y + z == c
    return (x, y, z)

File: src/Day1/part2.go
package Day1

import (
  "fmt"
  "strconv"
  "github.com/zylphrex/aoc2020/src/util"
)

func Day1Part2(input string) (string, error) {
  xs, err := util.ParseInts(input)
  if err != nil {
    return "", err
  }

  a, b, c, err := solveDay1Part2(2020, xs)
  return strconv.Itoa(a * b * c), err
}

func solveDay1Part2(t int, xs []int) (int, int, int, error) {
  for i, a := range xs {
    for j := i + 1; j < len(xs); j++ {
      b := xs[j]
      for k := j + 1; k < len(xs); k++ {
        c := xs[k]
        if a + b + c == t {
          return a, b, c, nil
        }
      }
    }
  }
  return 0, 0, 0, fmt.Errorf("No solution found")
}
